CMakeLists.txt

```diff
@@ -639,15 +639,15 @@ if(IREE_BUILD_PYTHON_BINDINGS)
   # "Bootstrapping" by first looking for the optional Development component
   # seems to be robust generally.
   # See: https://reviews.llvm.org/D118148
-  # If building Python packages, we have a hard requirement on 3.9+.
-  find_package(Python3 3.9 COMPONENTS Interpreter Development NumPy)
-  find_package(Python3 3.9 COMPONENTS Interpreter Development.Module NumPy REQUIRED)
+  # If building Python packages, we have a hard requirement on 3.10+.
+  find_package(Python3 3.10 COMPONENTS Interpreter Development NumPy)
+  find_package(Python3 3.10 COMPONENTS Interpreter Development.Module NumPy REQUIRED)
```

Comment:
I did a bit of research on how we've handled prior Python version updates and found these PRs:

* https://github.com/iree-org/iree/pull/14810
  * Note this comment: https://github.com/iree-org/iree/pull/14810#issuecomment-1693971471

    > If it were me, I would bump to 3.11 or 3.10. I see no value in the old versions for where we are. But if you are doing the work, I'm not going to push back.
* https://github.com/iree-org/iree/pull/14860

Here's some official guidance on dropping support in published packages: https://packaging.python.org/en/latest/guides/dropping-older-python-versions/

We could probably drop support for 3.9 in our next (next) release, 3.2.0? If we do that, we should first send out an announcement asking if anyone depends on it then also put together a policy on how long to support package versions. Looking through https://en.wikipedia.org/wiki/History_of_Python#Table_of_versions , we could choose some number of years since release or "end of full support". We could also survey related packages (tensorflow, torch, onnx, etc.) and use their minimum versions as a reference point.

Reply:
PyTorch required 3.9 or later (thus it is still "supported") but as Python 3.9 will be EOL 2025-10 I am perfectly fine with dropping it as part of our next release. Announcing this sounds good but I would say we could draft a policy at a later point and have this as a separate discussion (including the community of course).